### PR TITLE
libsql wal s3 backend

### PR DESCRIPTION
**Pull request by MarinPostma**
implement the S3 backend for AsyncStorage that stores segments on S3. Also add a small CLI to test libsql wal

### CODE REVIEW

File: libsql-wal/src/storage/async_storage.rs [hunk 1]
```diff
@@ -128,31 +154,62 @@ where
...
+        rcv.blocking_recv().unwrap()
     }
 }
 
-pub struct AsyncStorageInitConfig<S> {
-    storage: S,
-    max_in_flight_jobs: usize,
+pub struct AsyncStorageInitConfig<B> {
+    pub backend: Arc<B>,
+    pub max_in_flight_jobs: usize,
```
Comment (LucioFranco) [inline libsql-wal/src/storage/async_storage.rs:197]: can we avoid these pubs?

File: libsql-wal/src/storage/backend/s3.rs [hunk 2]
```diff
@@ -1,126 +1,471 @@
...
+        let body = SdkBody::retryable(move || {
+            let s = Self::new_inner(self.inner.clone());
+            SdkBody::from_body_1_x(s)
+        });
+
+        ByteStream::new(body)
+    }
+}
+
+impl<F> http_body::Body for FileStreamBody<F>
```
Comment (LucioFranco) [inline libsql-wal/src/storage/backend/s3.rs:416]: I don't totally follow why we need a custom body implementation? Might be worth adding a comment why here

**Review by LucioFranco — APPROVED** [review]
I have a few questions but approving to move this forward.

File: libsql-wal/src/storage/async_storage.rs (hunk 1 above)
Comment (MarinPostma) [reply libsql-wal/src/storage/async_storage.rs:197]: it's a config struct

File: libsql-wal/src/storage/backend/s3.rs (hunk 2 above)
Comment (MarinPostma) [reply libsql-wal/src/storage/backend/s3.rs:416]: I couldn't find how to wrap a stream, but I reckon we may have used https://docs.rs/stream-body/latest/stream_body/

**Pull request merged**